src/freenet/crypt/Hash.java

```diff
@@ -0,0 +1,205 @@
+/* This code is part of Freenet. It is distributed under the GNU General
+ * Public License, version 2 (or at your option any later version). See
+ * http://www.gnu.org/ for further details of the GPL. */
+package freenet.crypt;
+
+import java.nio.ByteBuffer;
+import java.security.MessageDigest;
+
+import net.i2p.util.NativeBigInteger;
+import freenet.support.HexUtil;
+
+/**
+ * The Hash class will generate the hash value of a given set of bytes and also verify that
+ * a hash matches a given set of bytes. The addBytes methods can be used to pass data into 
+ * a buffer that will be used to generate a hash. Once a hash is generated, the buffer is 
+ * cleared or reset. 
+ * @author unixninja92
+ *
+ */
+public final class Hash{
+	public static final HashType preferredHashAlg = HashType.SHA256;
+	private final HashType type;
+	private MessageDigest digest;
+	
+	/**
+	 * Creates an instance of Hash using the specified hashing algorithm.
+	 * @param type The hashing algorithm to use. 
+	 */
+	public Hash(HashType type){
+		this.type = type;
+		digest = type.get();
+	}
+	
+	/**
+	 * Generates the hash of all the bytes in the buffer added with the
+	 * addBytes methods. The buffer is then cleared after the hash has been
+	 * generated.
+	 * @return The generated hash of all the bytes added since last reset.
+	 */
+	public final byte[] genHash(){
+		byte[] result = digest.digest();
+		if(type == HashType.ED2K){
+			//ED2K does not reset after generating a digest. Work around this issue
+			digest.reset();
+		}else if(type == HashType.TTH){
+			//TTH's .reset method is broken or isn't implemented. Work around this bug
+			digest = type.get();
+		}
+		return result;
+	}
+	
+	/**
+	 * Generates the hash of only the specified bytes. The buffer is cleared before 
+	 * processing the input to ensure that no extra data is included. Once the hash
+	 * has been generated, the buffer is cleared again. 
+	 * @param input The bytes to hash
+	 * @return The generated hash of the data
+	 */
+	public final byte[] genHash(byte[]... input) {
+		digest.reset();
+		addBytes(input);
+		return genHash();
+	}
+	
+	/**
+	 * Generates the HashResult of all the bytes in the buffer added with the
+	 * addBytes methods. The buffer is then cleared after the hash has been
+	 * generated.
```

Comment:
With my experience with crypto APIs, it seems standard to have the coder keep track of internal state.